core/include/kernel/pm.h

```diff
@@ -0,0 +1,112 @@
+ * PM_STATE_HINT_POWER - When set device power shall be suspended/restored
+ * PM_STATE_HINT_IO - When set IO pins shall be suspended/restored
+ * PM_STATE_HINT_CONTEXT - When set the full context be suspended/restored
+ * PM_STATE_PLATFORM_MASK - Bit mask reserved for platform specific hints
+ */
+#define PM_STATE_HINT_CLOCK		BIT(0)
+#define PM_STATE_HINT_POWER		BIT(1)
+#define PM_STATE_HINT_IO		BIT(2)
+#define PM_STATE_HINT_CONTEXT		BIT(3)
+#define PM_STATE_PLATFORM_MASK		GENMASK_32(31, 16)
+
+/*
+ * PM_OP_SUSPEND: platform is suspending to a target low power state
+ * PM_OP_RESUME: platform is resuming from low power state
+ */
+enum pm_op {
+	PM_OP_SUSPEND = 0,
+	PM_OP_RESUME = 1,
+};
+
+/*
+ * Registered callbacks are called the ordering directives
+ * specified by the PM_CB_LEVEL_xxx value.
+ */
+enum pm_cb_level {
+	PM_CB_LEVEL_DRIVER = 0,
+	PM_CB_LEVEL_CORE_SERVICE,
+	PM_CB_LEVELS
+};
+
+#define PM_CB_HDL(_cb, _hdl, _level)		\
+		(struct pm_cb_hdl){		\
+			.callback = (_cb),	\
+			.handle = (_hdl),	\
+			.level = (_level),	\
+		}
+
+#define PM_CB_ARG_HANDLE(pm_cb_hdl)		((pm_cb_hdl)->handle)
+
+/*
+ * @callback - Registered callback function
+ * @handle - Registered private handler for the callback
+ * @level - Registered callback call order priority (PM_CB_PRIO_*)
+ */
+struct pm_cb_hdl {
+	/* Set by the caller when registering a callback */
+	int (*callback)(enum pm_op op, unsigned int pm_state,
+			const struct pm_cb_hdl *pm_cb_hdl);
+	void *handle;
+	uint8_t level;
+	/* Set by the system accoridng to execution context */
+	uint8_t flags;
+};
+
+/*
+ * Register a callback for suspend/resume sequence
+ *
+ * @pm_hdl: Reference callback registration directive
+ *
+ * Drivers and services can register a callback for the suspend and resume
+ * sequences and a private cookie passed as argument to the callback.
+ * The same callback is used for suspend and resume. First argument of the
+ * callback defines whether the device shall suspend or resume.
+ *
+ * Driver should tag its callback resource as unpaged (i.e KEEP_PAGER())
+ * since the callback is called from an unpaged execution context.
+ *
+ * Callback shall return 0 on success and return 1 in case of failure.
+ * This may be used during a failing suspend sequence to resume the
+ * suspended resources.
+ */
+void register_pm_cb(struct pm_cb_hdl *pm_hdl);
+
+/*
+ * Register a driver callback for generic suspend/resume
+ *
+ * @callback: Registered callback function
+ * @handle: Registered handle argument for the callback
+ */
+void register_pm_driver_cb(int (*callback)(enum pm_op op, unsigned int pm_state,
+					   const struct pm_cb_hdl *pm_cb_hdl),
+			   void *handle);
+
+/*
+ * Request call to registered PM callbacks
+ *
+ * @op: Either PM_OP_SUSPEND or PM_OP_RESUME
+ * @pm_state: Platform power state we are suspending to or resuming from
+ *
+ * Return 0 on success, 1 on failure of a callback.
+ */
+int pm_change_state(enum pm_op op, unsigned int pm_state);
```

Comment:
So now we have yet another set of return codes. Until this we had:
bool (true, false)
int (0 success, < 0 failure)
TEE_Result (TEE_SUCCES/0 success, anything else error)

Unless there's a good reason I think we should stick to TEE_Result just as the Linux kernel normally uses -errno.

Reply:
+1, since having various return codes is a pain in the end. We're already enforced to convert error codes to non TEE error codes in many cases on non-secure side when returning back from the TEE (KM/GK, Widevine, PKCS#11 to name a few). So trying to only use TEE_Result on secure side seems like a good idea.